target/snitch_cluster/sw/apps/snax-mac-mult/src/snax-mac-mult.c

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//----------------------------------
+// Author: Ryan Antonio (ryan.antonio@esat.kuleuven.be)
+// Description:
+// This test checks if a single Snitch core
+// Can control multiple accelerators attached to it.
+//
+// Note that this test encodes the CSRs in a manual way
+// because the write_csr and read_csr functions
+// found in sw/deps/riscv-opcodes/encoding.h
+// Do not allow variable inputs. The values need to be
+// explicitly stated due to the # stringitized arguments
+//----------------------------------
+
+#include "snrt.h"
+
+#include "data.h"
+#include "mac.h"
+
+int main() {
+    // Set err value for checking
+    int err = 0;
+
+    uint32_t *local_a_acc_0, *local_b_acc_0, *local_o_acc_0;
+    uint32_t *local_a_acc_1, *local_b_acc_1, *local_o_acc_1;
+    uint32_t *local_a_acc_2, *local_b_acc_2, *local_o_acc_2;
+    uint32_t *local_a_acc_3, *local_b_acc_3, *local_o_acc_3;
+
+    uint32_t SLICE_LEN = VEC_LEN / ACC_NUM;
+
+    // Calculate base addresses
+    local_a_acc_0 = (uint32_t *)snrt_l1_next();
+    local_b_acc_0 = local_a_acc_0 + VEC_LEN;
+    local_o_acc_0 = local_b_acc_0 + VEC_LEN;
+
+    local_a_acc_1 = local_a_acc_0 + SLICE_LEN;
+    local_b_acc_1 = local_b_acc_0 + SLICE_LEN;
+    local_o_acc_1 = local_o_acc_0 + SLICE_LEN;
+
+    local_a_acc_2 = local_a_acc_0 + 2 * SLICE_LEN;
+    local_b_acc_2 = local_b_acc_0 + 2 * SLICE_LEN;
+    local_o_acc_2 = local_o_acc_0 + 2 * SLICE_LEN;
+
+    local_a_acc_3 = local_a_acc_0 + 3 * SLICE_LEN;
+    local_b_acc_3 = local_b_acc_0 + 3 * SLICE_LEN;
+    local_o_acc_3 = local_o_acc_0 + 3 * SLICE_LEN;
+
+    // Manual declaration to optimize the part
+    // Where the compiler has to fetch the VEC_LEN
+    // From memory if we plug it directly into the
+    // csr_write functions. Adding this line
+    // pre-computes and does not need to do
+    // unnecessary lw or li instructions during csr setup
+
+    // Use data mover core to bring data from L3 to TCDM
+    if (snrt_is_dm_core()) {
+        size_t vector_size = VEC_LEN * sizeof(uint32_t);
+        snrt_dma_start_1d(local_a_acc_0, A, vector_size);
+        snrt_dma_start_1d(local_b_acc_0, B, vector_size);
+    }
+
+    // Wait for DMA transfer to finish
+    snrt_cluster_hw_barrier();
+
+    // Control all cores at this point
+    if (snrt_is_compute_core()) {
+        // Get start setup cycle
+        uint32_t tic_setup_mcycle = snrt_mcycle();
+
+        // We first set the 1st MAC accelerator's addresses
+        write_csr(976, (uint32_t)local_a_acc_0);
+        write_csr(977, (uint32_t)local_b_acc_0);
+        write_csr(979, (uint32_t)local_o_acc_0);
+
+        write_csr(980, 1);                 // Number of iterations
+        write_csr(981, SLICE_LEN);          // Vector length
+        write_csr(982, simple_mult_mode);  // Set simple multiplication
+
+        // Set for 2nd addresses
+        write_csr(1000, (uint32_t)local_a_acc_1);
```

Comment:
If you set all of these in a for loop, you can set an arbitrary number of accelerators, you can even unroll the loop if you really want to get the same code, you can even instruct the compiler with a pragma to fully unroll the loop?

Reply:
This is related to the problem above. I wanted to do it this elegantly, but that stupid `csrw` and `csrr` just doesn't work that way... I still haven't found an alternative sadly.